xarray/core/coordinates.py

```diff
@@ -106,9 +107,47 @@ def to_index(self, ordered_dims: Sequence[Hashable] = None) -> pd.Index:
             (dim,) = ordered_dims
             return self._data.get_index(dim)  # type: ignore
         else:
+            from pandas.core.arrays.categorical import factorize_from_iterable
+
             indexes = [self._data.get_index(k) for k in ordered_dims]  # type: ignore
-            names = list(ordered_dims)
-            return pd.MultiIndex.from_product(indexes, names=names)
+
+            # compute the sizes of the repeat and tile for the cartesian product
+            # (taken from pandas.core.reshape.util)
+            lenX = np.fromiter((len(index) for index in indexes), dtype=np.intp)
+            cumprodX = np.cumproduct(lenX)
+
+            if cumprodX[-1] != 0:
+                # sizes of the repeats
+                b = cumprodX[-1] / cumprodX
+            else:
+                # if any factor is empty, the cartesian product is empty
+                b = np.zeros_like(cumprodX)
+
+            # sizes of the tiles
+            a = np.roll(cumprodX, 1)
+            a[0] = 1
```

Comment:
Can we use more descriptive variables names here than `a` and `b`? E.g., maybe `repeat_counts` and `tile_sizes`?

Reply:
yes, done

Reply:
yes done